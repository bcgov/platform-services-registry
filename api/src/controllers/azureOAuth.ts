import axios from "axios";
import { logger } from "@bcgov/common-nodejs-utils";
// import { Response } from "express";
import qs from "qs";
import { Request, Response } from "express";

export const fetchAzureAccessToken = async (req: Request, res: Response): Promise<void> => {
  try {
    const options = {
      Host: "login.microsoftonline.com",
      client_id: `${process.env.AZURE_CLIENT_ID}`,
      scope: "https://graph.microsoft.com/.default",
      client_secret: process.env.AZURE_CLIENT_SECRET,
      grant_type: "client_credentials",
    };
    const url: string = `https://login.microsoftonline.com/${process.env.AZURE_TENANT_ID}/oauth2/v2.0/token`;
    axios
      .post(url, qs.stringify(options))
      .then((response) => {
        if (!response.data.access_token) {
          throw Error(
            "a result was returned from Microsoft's token endpoint, but it doesn't seem to have an access token included"
          );
        }
        res.status(200).json(response.data.access_token);
      })
      .catch((error) => {
        logger.info(`error: ${error}`);
        res.status(500);
      });
  } catch (err) {
    logger.error(`err = ${err.message}`);
  }
};

export default fetchAzureAccessToken;
