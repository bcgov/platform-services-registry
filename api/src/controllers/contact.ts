//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-15.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;

  const rv = validateObjProps(ContactModel.requiredFields, body);
  if (rv) {
    throw rv;
  }

  try {
    const result = await ContactModel.create(body);

    res.status(201).json(result);
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateContact = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { ContactModel } = dm;
  const { contactId } = params;
  const {
    firstName,
    lastName,
    email,
    githubId,
  } = body;

  try {
    const record = await ContactModel.findById(contactId);
    const aBody = {
      userId: record.userId,
      firstName,
      lastName,
      email,
      githubId,
      roleId: record.roleId,
    };

    const rv = validateObjProps(ContactModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ContactModel.update(contactId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update contact ID ${contactId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const requestContactEdit = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
