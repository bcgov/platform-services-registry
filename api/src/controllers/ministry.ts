//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from "@bcgov/common-nodejs-utils";
import { Request, Response } from "express";
import DataManager from "../db";
import shared from "../libs/shared";

const dm = new DataManager(shared.pgPool);

const fetchMinistrySponsors = async (
  req: Request,
  res: Response
): Promise<void> => {
  const { MinistryModel } = dm;
  try {
    const results = await MinistryModel.findAll();

    res.status(200).json(results);
  } catch (err) {
    const message = "Unable fetch ministries";
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export default fetchMinistrySponsors;
