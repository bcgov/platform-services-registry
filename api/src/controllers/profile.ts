//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import { generateNamespacePrefix } from '../db/utils';
import { AuthenticatedUser } from '../libs/authmware';
import shared from '../libs/shared';
import { isNotAuthorized, validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const uniqueNamespacePrefix = async (): Promise<string | undefined> => {
  const { ProfileModel } = dm;
  const attempts = 3;
  const candidates = Array.from({ length: attempts }, () => generateNamespacePrefix())
  const promises = candidates.map(c => ProfileModel.isNamespacePrefixUnique(c));

  try {
    const results = await Promise.all(promises);
    const values = results.map((cur, idx) => {
      if (cur) {
        return candidates[idx];
      }

      return;
    }).filter(v => typeof v !== 'undefined');

    return values.pop();
  } catch (err) {
    logger.error(err.message);
    return;
  }
}

export const fetchAllProjectProfiles = async (
  { user }: { user: AuthenticatedUser }, res: Response): Promise<void> => {
  const { ProfileModel } = dm;

  try {
    let results;
    if (user.roles.includes(USER_ROLES.ADMINISTRATOR)) {
      results = await ProfileModel.findAll();
    } else {
      results = await ProfileModel.findProfilesByUserId(user.id);
    }

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable fetch all project profiles';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    res.status(200).json(record);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch project profile with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const createProjectProfile = async (
  { body, user }: { body: any, user: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const data = { ...body, userId: user.id };

  // User cannot set the namespace prefix. If it exists, this
  // overwrites it with a place holder value. It will be replaced
  // with the actual value further on.
  const rv = validateObjProps(ProfileModel.requiredFields, {
    ...data,
    namespacePrefix: 'placeholder',
  });

  if (rv) {
    throw rv;
  }

  try {
    const namespacePrefix = await uniqueNamespacePrefix();

    if (!namespacePrefix) {
      throw errorWithCode(500, 'Unable to generate unique namespace prefix');
    }

    const results = await ProfileModel.create({ ...data, namespacePrefix });

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable create new project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProjectProfile = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const {
    name,
    description,
    categoryId,
    busOrgId,
    prioritySystem,
    notificationEmail,
    notificationSms,
    notificationMsTeams,
    paymentBambora,
    paymentPayBc,
    fileTransfer,
    fileStorage,
    geoMappingWeb,
    geoMappingLocation,
    schedulingCalendar,
    schedulingAppointments,
    idmSiteMinder,
    idmKeycloak,
    idmActiveDir,
    other,
  } = body;

  try {
    const record = await ProfileModel.findById(profileId);
    const aBody = {
      name,
      description,
      categoryId,
      busOrgId,
      prioritySystem,
      userId: record.userId,
      namespacePrefix: record.namespacePrefix,
      notificationEmail,
      notificationSms,
      notificationMsTeams,
      paymentBambora,
      paymentPayBc,
      fileTransfer,
      fileStorage,
      geoMappingWeb,
      geoMappingLocation,
      schedulingCalendar,
      schedulingAppointments,
      idmSiteMinder,
      idmKeycloak,
      idmActiveDir,
      other,
    };

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const rv = validateObjProps(ProfileModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ProfileModel.update(profileId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(profileId);

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    await ProfileModel.delete(profileId);

    res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = 'Unable to archive project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const addContactToProfile = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId, contactId } = params;

  try {
    await ProfileModel.addContactToProfile(Number(profileId), Number(contactId));

    res.status(202).end();
  } catch (err) {
    const message = `Unable to create contact`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileContacts = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ContactModel, ProfileModel } = dm;
  const { profileId } = params;

  try {
    const record = await ProfileModel.findById(Number(profileId));

    const notAuthorized = isNotAuthorized(record, user);

    if (notAuthorized) {
      throw notAuthorized;
    }

    const results = await ContactModel.findForProject(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch profile contacts with profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileEditRequests = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  try {
    res.status(201).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
