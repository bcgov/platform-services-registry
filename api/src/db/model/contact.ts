// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-06-22.
//

import { logger } from '@bcgov/common-nodejs-utils';
import { Pool } from 'pg';
import { CommonFields, Model } from './model';

export interface Contact extends CommonFields {
  firstName: string;
  lastName: string;
  email: string;
  githubId: string;
  roleId: number;
}

export default class ContactModel extends Model {
  table: string = 'contact';
  requiredFields: string[] = [
    'firstName',
    'lastName',
    'email',
    'roleId',
  ];
  pool: Pool;

  constructor(pool: any) {
    super();
    this.pool = pool;
  }

  async create(data: Contact): Promise<Contact> {
    const query = {
      text: `
        INSERT INTO ${this.table}
          (first_name, last_name, email, github_id, role_id)
          VALUES ($1, $2, $3, $4, $5) RETURNING *;`,
      values: [
        data.firstName,
        data.lastName,
        data.email,
        data.githubId.toLowerCase(),
        data.roleId,
      ],
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to create contact`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async findForProject(profileId: number): Promise<Contact[]> {
    const query = {
      text: `
        SELECT contact.*
          FROM contact
          JOIN profile_contact ON contact.id = profile_contact.contact_id
            WHERE profile_contact.profile_id = $1;
      `,
      values: [
        profileId,
      ],
    };

    try {
      return await this.runQuery(query);
    } catch (err) {
      const message = `Unable to fetch contacts for profile ${profileId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async update(contactId: number, data: Contact): Promise<Contact> {
    const values: any[] = [];
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            first_name = $1, last_name = $2, email = $3, github_id = $4,
            role_id = $5
          WHERE id = ${contactId}
          RETURNING *;`,
      values,
    };

    try {
      const record = await this.findById(contactId);
      const aData = { ...record, ...data };
      query.values = [
        aData.firstName,
        aData.lastName,
        aData.email,
        aData.githubId,
        aData.roleId,
      ];

      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to update contact ID ${contactId}`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }

  async delete(contactId: number): Promise<Contact> {
    const query = {
      text: `
        UPDATE ${this.table}
          SET
            archived = true
          WHERE id = ${contactId}
          RETURNING *;
      `,
    };

    try {
      const results = await this.runQuery(query);
      return results.pop();
    } catch (err) {
      const message = `Unable to archive contact`;
      logger.error(`${message}, err = ${err.message}`);

      throw err;
    }
  }
}
