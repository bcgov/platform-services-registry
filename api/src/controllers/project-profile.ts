//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import { USER_ROLES } from '../constants';
import DataManager from '../db';
import { generateNamespacePrefix } from '../db/utils';
import { AuthenticatedUser } from '../libs/authmware';
import { getAuthorization } from '../libs/authorization';
import { getDefaultCluster } from '../libs/profile';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const uniqueNamespacePrefix = async (): Promise<string | undefined> => {
  const { ProfileModel } = dm;
  const attempts = 3;
  const candidates = Array.from({ length: attempts }, () => generateNamespacePrefix())
  const promises = candidates.map(c => ProfileModel.isNamespacePrefixUnique(c));

  try {
    const results = await Promise.all(promises);
    const values = results.map((cur, idx) => {
      if (cur) {
        return candidates[idx];
      }

      return;
    }).filter(v => typeof v !== 'undefined');

    return values.pop();
  } catch (err) {
    logger.error(err.message);
    return;
  }
}

export const fetchAllProjectProfiles = async (
  { user }: { user: AuthenticatedUser }, res: Response): Promise<void> => {
  const { ProfileModel } = dm;

  try {
    let results;
    if (user.roles.includes(USER_ROLES.ADMINISTRATOR)) {
      results = await ProfileModel.findAll();
    } else {
      results = await ProfileModel.findProfilesByUserIdOrEmail(user.id, user.email);
    }

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable fetch all project profiles';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const projectDetails = await ProfileModel.findById(Number(profileId));

    const isAuthorized = getAuthorization(profileId, user, projectDetails);

    if (!(isAuthorized)) {
      throw isAuthorized;
    }

    res.status(200).json(projectDetails);
  } catch (err) {
    if (err.code) {
      throw err
    }

    const message = `Unable fetch project profile with ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const createProjectProfile = async (
  { body, user }: { body: any, user: any }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const data = { ...body, userId: user.id };

  // User cannot set the namespace prefix OR primary cluster name (current default cluster).
  // If it exists, this overwrites it with a place holder value. It will be replaced
  // with the actual value further on.
  const rv = validateObjProps(ProfileModel.requiredFields, {
    ...data,
    namespacePrefix: 'placeholder',
    primaryClusterName: 'placeholder',
  });

  if (rv) {
    throw rv;
  }

  try {
    const namespacePrefix = await uniqueNamespacePrefix();
    if (!namespacePrefix) {
      throw errorWithCode(500, 'Unable to generate unique namespace prefix');
    }

    const defaultCluster = await getDefaultCluster();
    if (!defaultCluster) {
      throw errorWithCode(500, 'Unable to set primary cluster id based on default cluster');
    }

    const results = await ProfileModel.create({ ...data, namespacePrefix, primaryClusterName: defaultCluster.name });

    res.status(200).json(results);
  } catch (err) {
    const message = 'Unable create new project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProjectProfile = async (
  { params, body, user }: { params: any, body: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;
  const {
    name,
    description,
    busOrgId,
    prioritySystem,
    notificationEmail,
    notificationSms,
    notificationMsTeams,
    paymentBambora,
    paymentPayBc,
    fileTransfer,
    fileStorage,
    geoMappingWeb,
    geoMappingLocation,
    schedulingCalendar,
    schedulingAppointments,
    idmSiteMinder,
    idmKeycloak,
    idmActiveDir,
    other,
    migratingLicenseplate,
  } = body;

  try {
    const currentProjectDetails = await ProfileModel.findById(profileId);
    const aBody = {
      name,
      description,
      busOrgId,
      prioritySystem,
      userId: currentProjectDetails.userId,
      namespacePrefix: currentProjectDetails.namespacePrefix,
      notificationEmail,
      notificationSms,
      notificationMsTeams,
      paymentBambora,
      paymentPayBc,
      fileTransfer,
      fileStorage,
      geoMappingWeb,
      geoMappingLocation,
      schedulingCalendar,
      schedulingAppointments,
      idmSiteMinder,
      idmKeycloak,
      idmActiveDir,
      other,
      primaryClusterName: currentProjectDetails.primaryClusterName,
      migratingLicenseplate,
    };

    const isAuthorized = getAuthorization(profileId, user, currentProjectDetails);

    if (!(isAuthorized)) {
      throw isAuthorized;
    }

    const rv = validateObjProps(ProfileModel.requiredFields, aBody);

    if (rv) {
      throw rv;
    }

    const results = await ProfileModel.update(profileId, aBody);

    res.status(200).json(results);
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProjectProfile = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  const { ProfileModel } = dm;
  const { profileId } = params;

  try {
    const isAuthorized = getAuthorization(profileId, user);

    if (!(isAuthorized)) {
      throw isAuthorized;
    }

    await ProfileModel.delete(profileId);

    res.status(204).end();
  } catch (err) {
    if (err.code) {
      throw err;
    }

    const message = 'Unable to archive project profile';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};
