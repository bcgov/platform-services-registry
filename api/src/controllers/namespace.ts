//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Created by Jason Leach on 2020-04-27.
//

'use strict';

import { errorWithCode, logger } from '@bcgov/common-nodejs-utils';
import { Response } from 'express';
import DataManager from '../db';
import { RequestEditType } from '../db/model/request';
import { AuthenticatedUser } from '../libs/authmware';
import { fulfillNamespaceEdit } from '../libs/fulfillment';
import { getNamespaceSet, getNamespaceSetQuotaOptions, mergeRequestedCNToNamespaceSet, validateQuotaRequestBody } from '../libs/quota-editing';
import shared from '../libs/shared';
import { validateObjProps } from '../libs/utils';

const dm = new DataManager(shared.pgPool);

export const createNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;
  const aBody = { ...body, profileId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    const results = await NamespaceModel.create(aBody);

    res.status(201).json(results);
  } catch (err) {
    const message = 'Unable create new namespace';
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespaces = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId } = params;

  try {
    const results = await NamespaceModel.findForProfile(Number(profileId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace for profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { namespaceId } = params;

  try {
    const results = await NamespaceModel.findById(Number(namespaceId));

    res.status(200).json(results);
  } catch (err) {
    const message = `Unable fetch namespace with ID ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const updateProfileNamespace = async (
  { params, body }: { params: any, body: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;
  const { name, clusterId } = body;
  const aBody = { name, profileId, clusterId };

  const rv = validateObjProps(NamespaceModel.requiredFields, aBody);
  if (rv) {
    throw rv;
  }

  try {
    await NamespaceModel.update(namespaceId, aBody);

    res.status(204).end();
  } catch (err) {
    const message = `Unable update project profile ID ${profileId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const archiveProfileNamespace = async (
  { params }: { params: any }, res: Response
): Promise<void> => {
  const { NamespaceModel } = dm;
  const { profileId, namespaceId } = params;

  try {
    await NamespaceModel.delete(namespaceId);

    res.status(204).end();
  } catch (err) {
    const message = `Unable to archive profile ${profileId} namespace ${namespaceId}`;
    logger.error(`${message}, err = ${err.message}`);

    throw errorWithCode(message, 500);
  }
};

export const fetchProfileQuotaOptions = async (
  { params, user }: { params: any, user: AuthenticatedUser }, res: Response
): Promise<void> => {
  try {
    // process request params to get namespaces
    const { profileId, clusterName } = params;
    const namespaces = await getNamespaceSet(profileId, user);
    if (!namespaces) {
      const errmsg = `Cant fetch namespaces for user ${user.id}`;
      throw new Error(errmsg);
    }

    // fetch quota options for namespaces with an optional cluster name
    const quotaOptions = await getNamespaceSetQuotaOptions(namespaces, clusterName);
    res.status(200).json(quotaOptions);
  } catch (err) {
    const message = `Unable to fetch quota options`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};

export const requestProfileQuotaEdit = async (
  { params, user, body }: { params: any, user: AuthenticatedUser, body: any }, res: Response
): Promise<void> => {
  const { RequestModel } = dm;
  try {
    // process request params to get namespaces
    const { profileId, clusterName } = params;
    const namespaces = await getNamespaceSet(profileId, user);
    if (!namespaces) {
      const errmsg = `Cant fetch namespaces for user ${user.id}`;
      throw new Error(errmsg);
    }

    // process request body for validation
    const quotaOptions = await getNamespaceSetQuotaOptions(namespaces, clusterName);
    if (!quotaOptions) {
      const errmsg = `Cant fetch quota options for user ${user.id}`;
      throw new Error(errmsg);
    }
    const rv = validateQuotaRequestBody(quotaOptions, body);
    if (rv) {
      throw rv;
    }

    // create a request record and send nats message
    const editObject = await mergeRequestedCNToNamespaceSet(body, namespaces, clusterName);
    if (!editObject) {
      const errmsg = 'Cant generate request edit object';
      throw new Error(errmsg);
    }
    const editType = RequestEditType.Namespaces;

    const { natsContext, natsSubject } = await fulfillNamespaceEdit(profileId, editType, editObject);
    await RequestModel.create({
      profileId,
      editType,
      editObject: JSON.stringify(editObject),
      natsSubject,
      natsContext: JSON.stringify(natsContext),
    })
    res.status(204).end();
  } catch (err) {
    const message = `Unable to update quota`;
    logger.error(`${message}, err = ${err.message}`);
    throw errorWithCode(message, 500);
  }
};
