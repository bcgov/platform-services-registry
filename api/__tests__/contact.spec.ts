//
// Copyright Â© 2020 Province of British Columbia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

import fs from 'fs';
import path from 'path';
import { Pool } from 'pg';
import { updateContact } from '../src/controllers/contact';
import FauxExpress from './src/fauxexpress';

const p0 = path.join(__dirname, 'fixtures/select-profile-contacts.json');
const selectContact = JSON.parse(fs.readFileSync(p0, 'utf8'));

const client = new Pool().connect();

describe('Contact event handlers', () => {
  let ex;

  beforeEach(() => {
    jest.clearAllMocks();
    // jest.resetAllMocks();

    ex = new FauxExpress();
  });

  it('A contact is updated', async () => {
    const body = {
      firstName: 'John',
      lastName: 'Doe',
      email: 'john@example.com',
      githubId: 'john1100',
    };
    const req = {
      params: { contactId: 233 },
      body,
    }

    client.query.mockResolvedValue({ rows: [selectContact[0]] });

    // @ts-ignore
    await updateContact(req, ex.res);

    expect(client.query.mock.calls).toMatchSnapshot();
    expect(ex.res.statusCode).toMatchSnapshot();
    expect(ex.responseData).toMatchSnapshot();
    expect(ex.res.status).toBeCalled();
    expect(ex.res.json).toBeCalled();
  });

  it('A contact fails to update', async () => {
    const body = {
      firstName: 'John',
      lastName: 'Doe',
      email: 'john@example.com',
      githubId: 'john1100',
    };
    const req = {
      params: { contactId: 233 },
      body,
    }

    client.query.mockImplementation(() => { throw new Error() });

    // @ts-ignore
    await expect(updateContact(req, ex.res)).rejects.toThrow();

    expect(ex.res.status).not.toBeCalled();
    expect(ex.res.json).not.toBeCalled();
  });
});
