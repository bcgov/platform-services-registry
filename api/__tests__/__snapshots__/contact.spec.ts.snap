// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact event handlers A contact is updated 1`] = `
Array [
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        233,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        SELECT * FROM contact
          WHERE id = $1 AND archived = false;",
      "values": Array [
        233,
      ],
    },
  ],
  Array [
    Object {
      "text": "
        UPDATE contact
          SET
            first_name = $1, last_name = $2, email = $3, github_id = $4,
            role_id = $5
          WHERE id = 233
          RETURNING *;",
      "values": Array [
        "John",
        "Doe",
        "john@example.com",
        "john1100",
        1,
      ],
    },
  ],
]
`;

exports[`Contact event handlers A contact is updated 2`] = `200`;

exports[`Contact event handlers A contact is updated 3`] = `
Object {
  "archived": false,
  "createdAt": "2020-09-10T18:14:13.436Z",
  "email": "jane@example.com",
  "firstName": "Jane",
  "githubId": "jane1100",
  "id": 233,
  "lastName": "Doe",
  "roleId": 1,
  "updatedAt": "2020-09-10T18:14:13.436Z",
}
`;
